pgmpy/base/AncestralBase.py

```diff
@@ -492,3 +527,85 @@ def get_reachable_nodes(self, node, u_type=None, v_type=None):
                 reachable.add(current)
                 queue.extend(self.get_neighbors(current, u_type=u_type, v_type=v_type))
         return reachable
+
+    def __eq__(self, other):
+        """
+        Checks if two MAGs are equal. Two MAGs are equal if they have the same
+        nodes, edges(including marks), latent variables, and variable roles
+
+        Parameters
+        ----------
+        other: MAG object
+            The other MAG to compare with
+
+        Returns
+        -------
+        bool
+            True if the MAGs are equal, False otherwise
+
+        Examples
+        --------
+        >>> from pgmpy.base import MAG
+        >>> mag1 = MAG(
+        ...     ebunch=[("X", "Y", "-", ">"), ("Y", "Z", "-", ">")],
+        ...     latents={"L"},
+        ...     roles={"exposure": "X"},
+        ... )
+        >>> mag2 = MAG(
+        ...     ebunch=[("X", "Y", "-", ">"), ("Y", "Z", "-", ">")],
+        ...     latents={"L"},
+        ...     roles={"exposure": "X"},
+        ... )
+        >>> mag1 == mag2
+        True
+
+        >>> mag3 = MAG(
+        ...     ebunch=[("X", "Y", "-", ">")], latents={"L"}, roles={"exposure": "X"}
+        ... )
+        >>> mag1 == mag3
+        False
+        """
+        if not isinstance(other, AncestralBase):
+            return False
+
+        self_edges = {
+            (u, v, frozenset(data["marks"].items()))
+            for u, v, data in self.edges(data=True)
+        }
+        other_edges = {
+            (u, v, frozenset(data["marks"].items()))
+            for u, v, data in other.edges(data=True)
+        }
+
+        return (
+            set(self.nodes()) == set(other.nodes())
+            and self_edges == other_edges
+            and self.latents == other.latents
+            and self.get_role_dict() == other.get_role_dict()
+        )
+
+    def copy(self):
+        """
+        Return a copy of the graph, preserving nodes, edges, marks, latents, and roles.
+
+        Returns
+        -------
+        AncestralBase
+            A new instance of the same class as self with all properties copied.
+        """
+        new_graph = self.__class__()
+
+        new_graph.add_nodes_from(self.nodes())
+
+        for u, v, data in self.edges(data=True):
+            u_mark, v_mark = data["marks"][u], data["marks"][v]
+            new_graph.add_edge(u, v, u_mark, v_mark)
+
+        if hasattr(self, "latents"):
```

Comment:
Why the `hasattr` check? Isn't it guaranteed that the object will have a `latents` attribute?